File: public/public-web/src/test/java/ch/difty/scipamato/publ/web/paper/browse/SimpleFilterPanelTest.java
package ch.difty.scipamato.publ.web.paper.browse;

import static org.assertj.core.api.Assertions.assertThat;

import org.apache.wicket.event.IEvent;
import org.apache.wicket.markup.html.form.FormComponent;
import org.apache.wicket.model.Model;
import org.jetbrains.annotations.NotNull;
import org.junit.jupiter.api.Test;

import ch.difty.scipamato.publ.entity.filter.PublicPaperFilter;
import ch.difty.scipamato.publ.web.common.PanelTest;

class SimpleFilterPanelTest extends PanelTest<SimpleFilterPanel> {

    private static final String PANEL = "panel";

    private static final int COMPONENTS_WITH_EVENT_HANDLER = 16;

    private int eventHandlerCallCount;

    @Override
    protected SimpleFilterPanel makePanel() {
        return new SimpleFilterPanel(PANEL, Model.of(new PublicPaperFilter()), "en");
    }

    @Override
    protected void assertSpecificComponents() {
        getTester().assertComponent(PANEL, SimpleFilterPanel.class);

        assertLabeledTextField(PANEL, "methodsSearch");
        assertLabeledTextField(PANEL, "authorsSearch");
        assertLabeledTextField(PANEL, "pubYearFrom");
        assertLabeledTextField(PANEL, "pubYearUntil");
        assertLabeledMultiSelect(PANEL, "populationCodes");
        assertLabeledMultiSelect(PANEL, "studyDesignCodes");
        assertLabeledMultiSelect(PANEL, "keywords");
        assertLabeledTextField(PANEL, "titleSearch");
    }

    private SimpleFilterPanel makePanelSpy() {
        return new SimpleFilterPanel(PANEL, Model.of(new PublicPaperFilter()), "en") {
            @Override
            void handleChangeEvent(@NotNull final IEvent<?> event, @NotNull final FormComponent component) {
                super.handleChangeEvent(event, component);
                eventHandlerCallCount++;
            }
        };
    }

    @Test
    void notChangingAnyField() {
        getTester().startComponentInPage(makePanelSpy());
        assertThat(eventHandlerCallCount).isEqualTo(0);
    }

    @Test
    void changingTextField() {
        getTester().startComponentInPage(makePanelSpy());
        getTester().executeAjaxEvent("panel:methodsSearch", "change");
        assertThat(eventHandlerCallCount).isEqualTo(COMPONENTS_WITH_EVENT_HANDLER);
    }

    @Test
    void changingMultiselectCombo() {
        getTester().startComponentInPage(makePanelSpy());
        getTester().executeAjaxEvent("panel:populationCodes", "change");
        assertThat(eventHandlerCallCount).isEqualTo(COMPONENTS_WITH_EVENT_HANDLER);
    }

    @Test
    void changingKeywordMultiselect() {
        getTester().startComponentInPage(makePanelSpy());
        getTester().executeAjaxEvent("panel:keywords", "change");
        assertThat(eventHandlerCallCount).isEqualTo(COMPONENTS_WITH_EVENT_HANDLER);
    }

}

File: public/public-web/src/test/kotlin/ch/difty/scipamato/publ/web/paper/browse/SimpleFilterPanelTest.kt
package ch.difty.scipamato.publ.web.paper.browse

import ch.difty.scipamato.publ.entity.filter.PublicPaperFilter
import ch.difty.scipamato.publ.web.common.PanelTest
import org.amshove.kluent.shouldBeEqualTo
import org.apache.wicket.event.IEvent
import org.apache.wicket.markup.html.form.FormComponent
import org.apache.wicket.model.Model
import org.junit.jupiter.api.Test

class SimpleFilterPanelTest : PanelTest<SimpleFilterPanel>() {

    private var eventHandlerCallCount = 0

    override fun makePanel(): SimpleFilterPanel =
        SimpleFilterPanel(PANEL, Model.of(PublicPaperFilter()), "en")

    override fun assertSpecificComponents() {
        tester.assertComponent(PANEL, SimpleFilterPanel::class.java)
        assertLabeledTextField(PANEL, "methodsSearch")
        assertLabeledTextField(PANEL, "authorsSearch")
        assertLabeledTextField(PANEL, "pubYearFrom")
        assertLabeledTextField(PANEL, "pubYearUntil")
        assertLabeledMultiSelect(PANEL, "populationCodes")
        assertLabeledMultiSelect(PANEL, "studyDesignCodes")
        assertLabeledMultiSelect(PANEL, "keywords")
        assertLabeledTextField(PANEL, "titleSearch")
    }

    private fun makePanelSpy(): SimpleFilterPanel = object :
        SimpleFilterPanel(PANEL, Model.of(PublicPaperFilter()), "en") {
        override fun handleChangeEvent(event: IEvent<*>, component: FormComponent<*>) {
            super.handleChangeEvent(event, component)
            eventHandlerCallCount++
        }
    }

    @Test
    fun notChangingAnyField() {
        tester.startComponentInPage(makePanelSpy())
        eventHandlerCallCount shouldBeEqualTo 0
    }

    @Test
    fun changingTextField() {
        tester.startComponentInPage(makePanelSpy())
        tester.executeAjaxEvent("panel:methodsSearch", "change")
        eventHandlerCallCount shouldBeEqualTo COMPONENTS_WITH_EVENT_HANDLER
    }

    @Test
    fun changingMultiselectCombo() {
        tester.startComponentInPage(makePanelSpy())
        tester.executeAjaxEvent("panel:populationCodes", "change")
        eventHandlerCallCount shouldBeEqualTo COMPONENTS_WITH_EVENT_HANDLER
    }

    @Test
    fun changingKeywordMultiselect() {
        tester.startComponentInPage(makePanelSpy())
        tester.executeAjaxEvent("panel:keywords", "change")
        eventHandlerCallCount shouldBeEqualTo COMPONENTS_WITH_EVENT_HANDLER
    }

    companion object {
        private const val PANEL = "panel"
        private const val COMPONENTS_WITH_EVENT_HANDLER = 16
    }
}
